File: src/test/java/uk/co/agware/carpet/TestFileTask.java
package uk.co.agware.carpet;

import org.junit.Assert;
import org.junit.Before;
import org.junit.Test;
import org.mockito.ArgumentCaptor;
import org.mockito.Mockito;
import uk.co.agware.carpet.change.tasks.FileTask;
import uk.co.agware.carpet.database.DatabaseConnector;

/**
 * Created by Philip Ward <Philip.Ward@agware.com> on 14/06/2016.
 */
public class TestFileTask {

    private DatabaseConnector databaseConnector;

    @Before
    public void buildMocks(){
        databaseConnector = Mockito.mock(DatabaseConnector.class);
        Mockito.when(databaseConnector.executeStatement(Mockito.anyString())).thenReturn(true);
    }

    @Test
    public void testAbsoluteFilePath(){
        FileTask task = new FileTask("Task 1", 1, "src/test/files/test.sql", ";");
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        ArgumentCaptor<String> statements = ArgumentCaptor.forClass(String.class);
        Mockito.verify(databaseConnector, Mockito.times(2)).executeStatement(statements.capture());
        Assert.assertEquals(2, statements.getAllValues().size());
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Table"));
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Other_Table"));
    }

    @Test
    public void testAbsoluteFilePathNoDelimiter(){
        FileTask task = new FileTask("Task 1", 1, "src/test/files/test.sql", "");
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        ArgumentCaptor<String> statements = ArgumentCaptor.forClass(String.class);
        Mockito.verify(databaseConnector, Mockito.times(2)).executeStatement(statements.capture());
        Assert.assertEquals(2, statements.getAllValues().size());
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Table"));
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Other_Table"));
    }

    @Test
    public void testAbsoluteFilePathNullDelimiter(){
        FileTask task = new FileTask("Task 1", 1, "src/test/files/test.sql", null);
        Assert.assertEquals("Task 1", task.getTaskName());
        Assert.assertEquals(1, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        ArgumentCaptor<String> statements = ArgumentCaptor.forClass(String.class);
        Mockito.verify(databaseConnector, Mockito.times(2)).executeStatement(statements.capture());
        Assert.assertEquals(2, statements.getAllValues().size());
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Table"));
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Other_Table"));
    }

    @Test
    public void testClasspathFile(){
        FileTask task = new FileTask("Task 2", 2, "classpath:classpathTest.sql", ","); // Testing a random delimiter as well
        Assert.assertEquals("Task 2", task.getTaskName());
        Assert.assertEquals(2, task.getTaskOrder());
        Assert.assertTrue(task.performTask(databaseConnector));
        ArgumentCaptor<String> statements = ArgumentCaptor.forClass(String.class);
        Mockito.verify(databaseConnector, Mockito.times(2)).executeStatement(statements.capture());
        Assert.assertEquals(2, statements.getAllValues().size());
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM Classpath"));
        Assert.assertTrue(statements.getAllValues().contains("SELECT * FROM class_path"));
    }
}

File: src/test/kotlin/uk/co/agware/carpet/TestFileTask.kt
package uk.co.agware.carpet

import com.nhaarman.mockito_kotlin.argumentCaptor
import com.nhaarman.mockito_kotlin.mock
import com.nhaarman.mockito_kotlin.times
import com.nhaarman.mockito_kotlin.verify
import org.jetbrains.spek.api.Spek
import org.jetbrains.spek.api.dsl.describe
import org.jetbrains.spek.api.dsl.given
import org.jetbrains.spek.api.dsl.it
import org.jetbrains.spek.api.dsl.on
import org.junit.platform.runner.JUnitPlatform
import org.junit.runner.RunWith
import uk.co.agware.carpet.change.tasks.FileTask
import uk.co.agware.carpet.database.DatabaseConnector
import uk.co.agware.carpet.exception.MagicCarpetParseException
import kotlin.test.assertEquals
import kotlin.test.assertFailsWith
import kotlin.test.assertTrue

@RunWith(JUnitPlatform::class)
class TestFileTask: Spek({

  describe("A FileTask Object") {

    var connection = mock<DatabaseConnector>()

    beforeEachTest {
      connection = mock<DatabaseConnector>()
    }

    given("A file on the classpath") {
      val task = FileTask("Test Task", 1, "classpath:classpathTest.sql", ",")

      on("performing the task") {

        task.performTask(connection)

        it("should execute each task statement") {
          val statements = argumentCaptor<String>()
          verify(connection, times(2))
                 .executeStatement(statements.capture())
          assertEquals(2, statements.allValues.size)
          assertTrue(statements.allValues.contains("SELECT * FROM Classpath"))
          assertTrue(statements.allValues.contains("SELECT * FROM class_path"))
        }
      }

    }

    given("A file system file") {
      val subject = FileTask("Test Task", 1, "src/test/files/test.sql")

      on("performing the task") {

        subject.performTask(connection)

        it("should execute each statement in the task") {
          val statements = argumentCaptor<String>()
          verify(connection, times(2))
                 .executeStatement(statements.capture())
          assertEquals(2, statements.allValues.size)
          assertTrue(statements.allValues.contains("SELECT * FROM Table"))
          assertTrue(statements.allValues.contains("SELECT * FROM Other_Table"))
        }
      }
    }

    it("should fail with a MagicCarpetParseException") {
      assertFailsWith<MagicCarpetParseException> {
        FileTask("A Failing Task", 2, "this/does/not/exist")
      }
    }

    it("should fail with a MagicCarpetParseException") {
      assertFailsWith<MagicCarpetParseException> {
        FileTask("Test Task", 1, "src/test/files/empty.sql")
      }
    }

    it("should fail with a MagicCarpetParseException") {
      assertFailsWith<MagicCarpetParseException> {
        FileTask("A Failing Task", 2, "classpath:this.does.not.exist")
      }
    }
  }
})
